Ally_Desktop_Assistant.py

```diff
@@ -0,0 +1,464 @@
+
+        elif "open google" in query:
+            speak("Opening Google...")
+            webbrowser.open("https://www.google.com/")
+
+        elif "open youtube" in query:
+            speak("Opening youtube...")
+            webbrowser.open("https://www.youtube.com/")
+
+
+
+        elif ((("what" in query) or ("tell" in query) or ("what's" in query)) and ("time" in query)) or (
+                query == "time"):
+            speak(f"It is {time_am_pm()}")
+
+        elif ((("what" in query) or ("tell" in query) or ("what's" in query) or ("today's" in query)) and (
+                "date" in query)) or (query == "date"):
+            speak(f"It is {get_date()}")
+
+        elif (("what is your name" in query) or (("tell" in query) and "your name" in query) or (
+                "who are you" in query) or (("what's" in query) and ("name" in query))) or query == name.lower():
+            speak(f"I am {name}. I am your digital assistant. Please tell me how may I help you...")
+
+        elif ((query == "hi") or (query == "hello")) and (name.lower() in query):
+            speak(f"I am {name}. I am your assistant. Tell me how can I help you...")
+
+        elif ("play" in query) and (("music" in query) or ("song" in query)):
+            play_music()
+
+        elif (('close' in query) or('quit' in query) or ('exit' in query)) and (('yourself' in query) or (name.lower() in query)):
+            exit_it()
+
+        else:
+            speak("Searching on Google...")
+            search_google(query)
+
+
+def voice_command_query(event):
+    query = take_command()
+    if not (query is None) or (query == ""):
+        show(f"You : {query}")
+        start_assistant(query)
+
+
+def written_command_query(event):
+    query = command_var.get()
+    query = query.strip()
+    command_pad.delete(0, END)
+    if not (query == ""):
+        show(f"You : {query}")
+        start_assistant(query)
+    else:
+        pass
+
+
+def give_space(event):
+    text_entered = command_var.get()
+    try:
+        if text_entered[0] != ' ':
+            command_pad.insert(0, ' ')
+    except:
+        pass
+
+
+def exit_it():
+    root.destroy()
+
+
+def add_music_path():
+    music_path = rf"{askdirectory()}"
+
+    try:
+        with open(data_filename) as file_obj:
+            data_dict = json.load(file_obj)
+            data_dict["music_folder_path"] = music_path
+
+        with open(data_filename, 'w') as f_obj:
+            json.dump(data_dict, f_obj)
+
+    except:
+        data_dict = {"voice": "Female", "music_folder_path": music_path}
+        with open(data_filename, 'w') as f_obj:
+            json.dump(data_dict, f_obj)
+
+    tmsg.showinfo('Path Added/Changed', f'Music path added/changed.\n\nCurrent Music Path\n{music_path}')
+
+
+def help_user():
+    tmsg.showinfo("Help", f"This is a desktop assisant named {name}.\nYou can use it to make your work easy.\nTo know more about it, watch the tutorial.")
+
+
+def open_tut():
```

Comment:
Rename the method `open_in_browser(url)` and make it generic. Replace with it wherever ` webbrowser.open` is used